SafetyGraphics/safetyGraphics
# Add Chart Code export

# Overview

This PR greatly simplifies the charting workflow in safetyGraphics by moving details from the charting module (`mod_chartTab.R`) and putting them in to the chart configuration object (`makeChartConfig.R` and friends). This change greatly simplifies the export workflow by allowing many components of the chart objects to be reused. 

# Details

See the posts below for a longer discussion of the approach used here. Here's a quick list of changes: 

- Moved chart-level parameter management from `mod_chartTab.R ` to the new `makeChartParams.R`. This allows for re-use in the export workflow. 
- Added `makeChartExport.R` to create re-usable R scripts for charts/mappings.
- Added `makeChartConfigFunctions.R` to deal with binding functions to chart objects including the `ui`, `server` and `main` functions discussed below. 
- The updates `makeChartConfigFunctions.R` allowed for a major simplification of `mod_chartTab.R` including the deletion of 
the `mod_chartsRenderXXX.R` family of functions. 
- Replaced the reports tab with "Download html report" and "Download R script" buttons in the headers for each chart (implemented in `mod_chartTab.R`)
- Rewrote `inst/report/safetyGraphicsReport.Rmd` using the new framework. Biggest change is that the report only supports one chart (instead of multiple).  
 
Fixes #518 #595

# Test Notes

Checkout the branch and use the following script to initialize the app with the test modules and static plots included:

```
charts<-makeChartConfig()
charts[["rtf_render"]]$order<- -1
charts[["safetyOutlierExplorerModule"]]$order<-1
charts[["safetyOutlierExplorerStatic"]]$order<-1
safetyGraphicsApp(charts=charts)
```

Then test the following:

- Do a general review of the charting tab, and make sure that everything (filtering, mapping, etc) is working as expected. 
- Export the html reports for each chart and confirm that they work as expected. 
- Export the R script for each chart and confirm that running the script draws the chart. Note that no changes to the script are needed if you are using the default data, but the script needs updates when using custom data.
